### PR TITLE
refactor(server): extract add/remove assets logic to utility function

### PR DESCRIPTION
**Pull request by danieldietzler**

### CODE REVIEW

**Review by jrasm91 — APPROVED** [review]
Looks really nice! Thanks!

File: server/src/cores/access.core.ts [hunk 1]
```diff
@@ -87,7 +87,7 @@ export class AccessCore {
   async checkAccess(auth: AuthDto, permission: Permission, ids: Set<string> | string[]) {
     const idSet = Array.isArray(ids) ? new Set(ids) : ids;
     if (idSet.size === 0) {
-      return new Set();
+      return new Set<string>();
```
Comment (jrasm91) [inline server/src/cores/access.core.ts:90]: Can you revert this change?

File: server/src/repositories/album.repository.ts [hunk 2]
```diff
@@ -261,7 +255,7 @@ export class AlbumRepository implements IAlbumRepository {
   }
 
   @GenerateSql({ params: [{ albumId: DummyValue.UUID, assetIds: [DummyValue.UUID] }] })
-  async addAssets({ albumId, assetIds }: AlbumAssets): Promise<void> {
+  async addAssetIds(albumId: string, assetIds: string[]): Promise<void> {
```
Comment (jrasm91) [inline server/src/repositories/album.repository.ts:258]: You need to change the decorator to match

File: server/src/services/album.service.ts [hunk 3]
```diff
@@ -164,33 +164,16 @@ export class AlbumService {
...
-    }
+    const results = await addAssets(
+      auth,
+      { accessRepository: this.accessRepository, repository: this.albumRepository },
+      { id, assetIds: dto.ids },
+    );
 
     const newAssetIds = results.filter(({ success }) => success).map(({ id }) => id);
     if (newAssetIds.length > 0) {
-      await this.albumRepository.addAssets({ albumId: id, assetIds: newAssetIds });
```
Comment (jrasm91) [inline server/src/services/album.service.ts:193]: It might be better to use find instead the way it would abort as soon as possible instead of going through the whole list.

File: server/src/services/album.service.ts [hunk 4]
```diff
@@ -206,31 +189,14 @@ export class AlbumService {
...
-    }
+    const results = await removeAssets(
+      auth,
+      { accessRepository: this.accessRepository, repository: this.albumRepository },
+      { id, assetIds: dto.ids, permissions: [Permission.ASSET_SHARE, Permission.ALBUM_REMOVE_ASSET] },
+    );
 
     const removedIds = results.filter(({ success }) => success).map(({ id }) => id);
     if (removedIds.length > 0) {
-      await this.albumRepository.removeAssets(id, removedIds);
```
Comment (jrasm91) [inline server/src/services/album.service.ts:233]: Same here

File: server/src/repositories/album.repository.ts [hunk 5]
```diff
@@ -203,7 +197,7 @@ export class AlbumRepository implements IAlbumRepository {
 
   @GenerateSql({ params: [DummyValue.UUID, [DummyValue.UUID]] })
   @Chunked({ paramIndex: 1 })
-  async removeAssets(albumId: string, assetIds: string[]): Promise<void> {
```
Comment (mertalev) [inline server/src/repositories/album.repository.ts:206]: Nit: I think I like `addAssets`/`removeAssets` better than `addAssetIds`/`removeAssetIds`.

File: server/src/utils/asset.util.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+      results.push({ id: assetId, success: false, error: BulkIdErrorReason.NO_PERMISSION });
+      continue;
+    }
+
+    results.push({ id: assetId, success: true });
+  }
+
+  const newAssetIds = results.filter(({ success }) => success).map(({ id }) => id);
+  if (newAssetIds.length > 0) {
+    await repository.addAssetIds(dto.id, dto.assetIds);
```
Comment (jrasm91) [inline server/src/utils/asset.util.ts:44]: Is this a bug? Surely it should be the new asset IDs.

File: server/src/utils/asset.util.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  addAssetIds: (id: string, assetIds: string[]) => Promise<void>;
+  removeAssetIds: (id: string, assetIds: string[]) => Promise<void>;
+}
+
+export const addAssets = async (
+  auth: AuthDto,
+  repositories: { accessRepository: IAccessRepository; repository: IBulkAsset },
+  dto: { id: string; assetIds: string[] },
+) => {
+  const { accessRepository, repository } = repositories;
```
Comment (mertalev) [inline server/src/utils/asset.util.ts:18]: Nit: you can destructure the parameter directly.

File: server/src/utils/asset.util.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+  const newAssetIds = results.filter(({ success }) => success).map(({ id }) => id);
+  if (newAssetIds.length > 0) {
+    await repository.addAssetIds(dto.id, dto.assetIds);
+  }
+
+  return results;
+};
+
+export const removeAssets = async (
```
Comment (mertalev) [inline server/src/utils/asset.util.ts:50]: I imagine these set operations have the potential to get expensive with huge libraries. Albums will generally have manageable sizes, but if we intend to reuse this elsewhere, we should be mindful of how many assets this might be handling and how often it might get called.

**Review by mertalev — APPROVED** [review]
LGTM!

File: server/src/utils/asset.util.ts (hunk 6 above)
Comment (mertalev) [reply server/src/utils/asset.util.ts:44]: Ooh, great catch

File: server/src/repositories/album.repository.ts (hunk 5 above)
Comment (jrasm91) [reply server/src/repositories/album.repository.ts:206]: There is another one - get assets - which returns IDs. All three methods operate on IDs so it is a little weird to have them names differently imo.

Comment (mertalev) [reply server/src/repositories/album.repository.ts:206]: Ah, fair. `getAssetIds` is definitely better to distinguish between fetching the ID and the entity, and it's better for them to have consistent naming in that case.

Comment (jrasm91) [reply server/src/repositories/album.repository.ts:206]: Yeah that was my thought process as well.

File: server/src/utils/asset.util.ts (hunk 7 above)
Comment (danieldietzler) [reply server/src/utils/asset.util.ts:18]: I know, but I didn't want to cluster the parameters even more lol

File: server/src/utils/asset.util.ts (hunk 8 above)
Comment (danieldietzler) [reply server/src/utils/asset.util.ts:50]: I think - at least in the near future - the only use cases are stacks and memories. Both shouldn't be too large, right?

File: server/src/utils/asset.util.ts (hunk 6 above)
Comment (danieldietzler) [reply server/src/utils/asset.util.ts:44]: Dang you're good

File: server/src/utils/asset.util.ts (hunk 7 above)
Comment (danieldietzler) [reply server/src/utils/asset.util.ts:18]: You would prefer destructuring them directly?

Comment (mertalev) [reply server/src/utils/asset.util.ts:18]: Personal preference! I only destructure at the top of a function body if doing it in the parameter will blow it up to multiple lines.
